### PR TITLE
[CssSelector] Added cache on top of CssSelectorConverter

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/CssSelector/CssSelectorConverter.php [hunk 1]
```diff
@@ -57,6 +62,19 @@ public function __construct(bool $html = true)
...
+
+            return self::$htmlCache[$prefix][$cssExpr] = $this->translator->cssToXPath($cssExpr, $prefix);
+        }
+
+        if (self::$xmlCache[$prefix][$cssExpr] ?? false) {
+            return self::$xmlCache[$prefix][$cssExpr];
+        }
+
+        return self::$xmlCache[$prefix][$cssExpr] = $this->translator->cssToXPath($cssExpr, $prefix);
+
```
Comment (OskarStark) [inline src/Symfony/Component/CssSelector/CssSelectorConverter.php]: Blank line should be removed

**Comment by stof** [comment]
should we use 2 static properties for the cache, or an instance property ?

**Comment by stof** [comment]
Also, the code may be simpler if we define an instance-level `cache` property, being a reference to either `self::$xmlCache` or `self::$htmlCache`. This way, only the constructor needs to know about the 2 separate caches.

**Comment by lyrixx** [comment]
A New crawler is instanced on each method call (for the results). Then each time you call filter() a new css selector converter is instansiated. That's why the cache is static 

I will implement your last idea 👍

File: src/Symfony/Component/CssSelector/CssSelectorConverter.php [hunk 2]
```diff
@@ -57,6 +64,10 @@ public function __construct(bool $html = true)
      */
     public function toXPath(string $cssExpr, string $prefix = 'descendant-or-self::')
     {
-        return $this->translator->cssToXPath($cssExpr, $prefix);
+        if ($this->cache[$prefix][$cssExpr] ?? false) {
+            return $this->cache[$prefix][$cssExpr];
+        }
+
+        return $this->cache[$prefix][$cssExpr] = $this->translator->cssToXPath($cssExpr, $prefix);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/CssSelector/CssSelectorConverter.php]: would be slightly faster, we use this style elsewhere:
`return $this->cache[$prefix][$cssExpr] ?? $this->cache[$prefix][$cssExpr] = $this->translator->cssToXPath($cssExpr, $prefix);`

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/CssSelector/CssSelectorConverter.php (hunk 2 above)
Comment (lyrixx) [reply src/Symfony/Component/CssSelector/CssSelectorConverter.php]: Fixed thanks. I also added a bullet point in the CHANGELOG

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Pull request merged**